### PR TITLE
chore: validate headers as per http spec

### PR DESCRIPTION
**Pull request by amitksingh1490**

### CODE REVIEW

**Comment by kyri-petrou** [comment]
Hey guys, before removing us from the benchmark because of a missing Date header (which is because zio-http doesn't provide a way currently to add that), do you want to maybe [update tailcall to the latest version first](https://github.com/tailcallhq/graphql-benchmarks/pull/157)?

**Comment by kyri-petrou** [comment]
Also, given that this is a graphql library benchmark and not an HTTP server benchmark, perhaps this is the spec you should be checking against: https://graphql-http.com/

**Comment by amitksingh1490** [comment]
@kyri-petrou we are not removing anyone just adding a test for future proofing.

Tailcall version update we will do with npm so that we get rennovate updates.

This PR we will only merge once everyone adhere to the Http spec.

**Comment by kyri-petrou** [comment]
Here's a list of things you need to fix by the way just to save you the trouble of setting things up:

image

**Comment by kyri-petrou** [comment]
@amitksingh1490 look, we got off the wrong foot on this one. The reason this feels like a cheap shot is because if we were to follow both the HTTP and GraphQL-over-HTTP spec to the letter, we'd find non-compliance in more than 1 libraries. However, your knowledge of `zio-http` makes this feel like a targeted attempt given that this PR came out just a few hours after the new release of Caliban which is outperforming tailcall

Furthermore, to echo your argument above, is it fair that Caliban is fully compliant with the graphql-over-http spec and tailcall (and perhaps others) are not?

Having said all that, Caliban is not an HTTP server library, it's a GraphQL library. We provide server integrations for UX reasons, but we don't control the server implementation itself. Since it's getting late here, _tomorrow_ I'll open an issue in `zio-http` about this to have it added

**Comment by amitksingh1490** [comment]
@kyri-petrou 
>Furthermore, to echo your argument above, is it fair that Caliban is fully compliant with the graphql-over-http spec and tailcall (and perhaps others) are not?

Its unfair thats why I said, We can add a check for graphql-over-http spec also.

>we'd find non-compliance in more than 1 libraries.

I will fix them, they are open source whats the problem in that.


The discussion should be around how can we make benchmarks fair. Instead of finger pointing.

I created a PR because this is a known problem with graphQL over http benchmarks and I knew about it and thats the first thing I checked. To make sure that the servers are doing the same amount of work and to keep the benchmark around the graphQL implementations.
https://github.com/graphql-crystal/benchmarks/issues/234.

I have myself spent 2 years implementing and optimising ZIO-HTTP and used caliban in my previous orgs. I love those libs. I have no intention to remove it from the benchmark.

I think the ideal way should be we verify the specs both `http` and `graphql-over-http`. And results show them with report. I was going to that in my PR and ask review what other think about it.

I was even going to open a discussion whats the best way to do it. using this PR as reference but before I could do that all of this happened.

I'm open to discussing this in detail, but please ensure the conversation remains calm and respectful.

**Comment by tusharmath** [comment]
@kyri-petrou Benchmarks gets the best of us, doesn't it?

I love and have high regards for what you guys are doing in Caliban. In fact Tailcall was built on ZIO HTTP and Caliban before transitioning to Rust. We realised quickly that performance on the JVM isn't enough for our use case and decided to rewrite everything.

The date header is important piece of the test, because it includes a sys call and memory allocation and changes rapidly. Something that the JVM is not optimised for. 

Here is the thing, and I wish to confess publicly the "cheap shot" - Server date header is required as per the http spec, and all web servers by default produce this header except ZIO HTTP. This was a conscious decision taken on my part because I never felt the need of this header and it caused massive GC thrashing on the JVM. If someone wanted to be http spec compliant then they could add the date header in the response. We enabled the header only in case of public benchmarks having a mandate of a date header.


Unfortunately if you enable this header you will drop in performance, perhaps significantly and that bothers me as much as it bothers you.


Unlike the GraphQL over Http, the http spec is a well established standard and must be obeyed. In my humble opinion the right thing to do is to enable the date header in zio http by default, but I don't think I am the right person to do it. Feel free to raise a PR and I would be happy to merge it. 

Once again, apologies if this PR has offended or belittled your efforts in any way. Our objective is to be reasonable and open for public scrutiny.

**Comment by kyri-petrou** [comment]
@tusharmath I've opened a PR in `zio-http` to add the `Date` header to responses by default: https://github.com/zio/zio-http/pull/2773. We'll need to wait until that makes it into a release though

**Comment by tusharmath** [comment]
@kyri-petrou Awesome! I think we had this feature sometime back, not sure when it got removed. I can see it commented out everywhere in the examples.

**Comment by tusharmath** [comment]
@kyri-petrou I have merged the PR, Do you want to try it out with the snapshot version once?

**Comment by kyri-petrou** [comment]
> @kyri-petrou I have merged the PR, Do you want to try it out with the snapshot version once?

@tusharmath Sure thing, although I don't think we should merge a PR with a dependency on a snapshot version of `zio-http`, since `zio-http` breaks binary compatibility between RC versions and we usually have to create a new Caliban release to resolve them

**Comment by tusharmath** [comment]
We could do a snapshot release?

**Comment by kyri-petrou** [comment]
> We could do a snapshot release for Caliban also?

The way that Caliba's CI works, we can only do snapshot releases when merging into master. And we don't want to have our master branch depend on a snapshot version of `zio-http`.

By the way since RC6 was recently released, I _think_ the current zio-http snapshot should still be binary compatible

**Comment by tusharmath** [comment]
@kyri-petrou Tracking your proposed changes here — https://github.com/tailcallhq/tailcall/issues/1766

**Comment by kyri-petrou** [comment]
@tusharmath nice :) By the way I opened a draft PR with the latest zio-http version (which has the Date header added by default).

Although I also updated the zio version to the latest 2.1.x RC one as I wanted to check how is it performing against the 2.0.x version in a more realistic benchmarking setup

**Comment by tusharmath** [comment]
What did you observe?

File: run_benchmarks.sh [hunk 1]
```diff
@@ -12,56 +12,82 @@ function killServerOnPort() {
...
+validate_date_change() {
+  local first_date=$(fetch_date)
+  echo "First Date: $first_date"
+  if [ -z "$first_date" ]; then
+    echo "Error: Date header is missing in the response."
+    return 0 # Failure
+  fi
 
-    bash "test_query.sh"
+  sleep 1
```
Comment (amitksingh1490) [inline run_benchmarks.sh:39]:
```suggestion
  sleep 2
```

**Pull request closed without merging**